packages/core/src/browser/tree/tree.ts

```diff
@@ -204,6 +208,17 @@ export class TreeImpl implements Tree {
         return id !== undefined ? this.nodes[id] : undefined;
     }
 
+    getNodes(test: (node: TreeNode) => boolean): TreeNode[] {
```

Comment:
Looks like very expansive operation. Is there a way to replace clients with `getNode(id)`  by redesigning their tree structure?

Reply:
Then what `id` would you like to use ? Two different nodes can hold the same file uri (which was the old id).

Reply:
`getNode(id)` function is still in the interface. `getNodes()` is only called when there is a need. Having multiple roots opened in the file navigator means there could be more than one file associated with the same uri (like the scenario that @marechal-p described above).

Reply:
Suppose there is `file://foo/bar/baz/x.txt` file, and 2 workspaces `file://foo` and `file://foo/bar`.

If a file tree node have an URI of form `${workspaceUri}:${fileUri}` or something like that (so `file://foo:file://foo/bar/baz/x.txt` and `file://foo/bar:file://foo/bar/baz/x.txt`), then when a file is changed one can take a list of all workspaces, loop over them to create node ids and call `getNode` for each.

I assume that a list of workspaces will be short in most of cases (i have maximum 10 at the time), in comparison the file tree only for `node_modules` folder has over 800 children.

Reply:
> When this happens we have no context of this x.txt and therefore would have no idea which node to highlight in the file tree.

I don't understand how `getNodes()` makes difference? We always know workspace roots, so it is not an issue to create `file://foo:file://foo/bar/baz/x.txt` and  `file://foo/bar:file://foo/bar/baz/x.txt` and call `getNode` for them.

Also sorry i meant `${workspaceRootUri}:${fileUri}`, not `${workspaceUri}:...`

Reply:
Concatenating the absolute root uris with the absolute file uris seems to be the most simple way to craft an id indeed. The `getNodes` was to make the loop you talked about part of the implementation, but thinking back it doesn't make really much sense for the `Tree` interface indeed.

Reply:
@akosyakov fixed